include/swift/AST/Evaluator.h

```diff
@@ -0,0 +1,264 @@
+///
+///            Cache the given result.
+///
+///         bool isInFlight() const;
+///
+///           Whether we're already computing this value.
+///
+///         void setInFlight() const;
+///
+///           Indicate that we're about to compute this value. After this,
+///           \c isInFlight() should return \c true until \c cacheResult()
+///           is called.
+class Evaluator {
+  /// The diagnostics engine through which any cyclic-dependency
+  /// diagnostics will be emitted.
+  DiagnosticEngine &diags;
+
+  /// A vector containing all of the active evaluation requests, which
+  /// is treated as a stack.
+  std::vector<AnyRequest> activeRequests;
+
+  /// A cache that stores the results of requests.
+  llvm::DenseMap<AnyRequest, Optional<AnyValue>> cache;
+
+public:
+  /// Construct a new evaluator that can emit cyclic-dependency
+  /// diagnostics through the given diagnostics engine.
+  Evaluator(DiagnosticEngine &diags);
+
+  /// Evaluate the given request and produce its result,
+  /// consulting/populating the cache as required.
+  template<typename Request>
+  typename Request::OutputType operator()(const Request &request) {
+    return getResult(request);
+  }
+
+  /// Evaluate a set of requests and return their results as a tuple.
+  ///
+  /// Use this to describe cases where there are multiple (known)
+  /// requests that all need to be satisfied.
+  template<typename ...Requests>
+  std::tuple<typename Requests::OutputType...>
+  operator()(const Requests &...requests) {
+    return std::tuple<typename Requests::OutputType...>((*this)(requests)...);
+  }
+
+private:
+  /// Diagnose a cycle detected in the evaluation of the given
+  /// request.
+  void diagnoseCycle(const AnyRequest &request);
+
+  /// Retrieve the result of produced by evaluating a request that can
+  /// be cached.
+  template<typename Request,
+           typename std::enable_if<Request::isEverCached>::type * = nullptr>
+  typename Request::OutputType getResult(const Request &request) {
+    // The request can be cached, but check a predicate to determine
+    // whether this particular instance is cached. This allows more
+    // fine-grained control over which instances get cache.
+    if (request.isCached())
+      return getResultCached(request);
+
+    return getResultUncached(request);
+  }
+
+  /// Retrieve the result of produced by evaluating a request that
+  /// will never be cached.
+  template<typename Request,
+           typename std::enable_if<!Request::isEverCached>::type * = nullptr>
+  typename Request::OutputType getResult(const Request &request) {
+    return getResultUncached(request);
+  }
+
+  /// Produce the result of the request without caching.
+  template<typename Request>
+  typename Request::OutputType getResultUncached(const Request &request) {
+    // Push this request onto the active-requests stack while we
+    // evaluate it.
+    activeRequests.push_back(request);
```

Comment:
Why push/pop it if you can't / aren't going to detect cycles involving it?

Reply:
I'm diagnosing cycles based on the stack, but you're right (per above) that I should use this for cycle detection as well.